File: scripts/px4_cmd/vehicle_external_command.py
# -*- coding: utf-8 -*-
# Copyright (c) 2023 易鹏 中山大学航空航天学院
# Copyright (c) 2023 Peng Yi, Sun Yat-Sen University, School of Aeronautics and Astronautics

import time
import rospy
import threading
from typing import List, Optional
from tf.transformations import euler_from_quaternion
from px4_cmd.msg import Command
from px4_cmd.custom_command import *
from geometry_msgs.msg import PoseStamped
from geometry_msgs.msg import TwistStamped

PI = 3.14159265358979323846

class vehicle_external_command:
    def __init__(self) -> None:
        self.update_time: float = 0.02
        self.init_x: float = 0
        self.init_y: float = 0
        self.init_z: float = 0
        self.init_R: float = 0
        self.init_P: float = 0
        self.init_Y: float = 0
        self.external_cmd: Command = Command()
        self.position: List[float] = [0.0, 0.0, 0.0]
        self.attitude: List[float] = [0.0, 0.0, 0.0]
        self.velocity: List[float] = [0.0, 0.0, 0.0]
        self.angle_rate: List[float] = [0.0, 0.0, 0.0]
        self.quaternion: List[float] = [0.0, 0.0, 0.0, 0.0]
        self.pos_cb_mutex = threading.Lock()
        self.vel_cb_mutex = threading.Lock()
        self.change_cmd_mutex = threading.Lock()
        pass

    def start(self, node: Optional[str] = None) -> None:
        """start API node

        Args:
            node (str | None): node name for vehicle, defined in topic name: /{node}/mavros/.... if None means single vehicle simulation
        """        
        node_name = node
        if node_name == None:
            node_name: str = ""
            topics = rospy.get_published_topics()
            for topic in topics:
                if "/mavros" in topic[0]:
                    node_name = topic[0].split("/mavros")[0].strip('/').strip()
                    break
        # start external command thread
        topic_header = "/" + node_name + "/mavros/"
        self.external_cmd.Mode = Command.Move
        self.external_cmd.Move_frame = Command.ENU
        self.external_cmd.Move_mode = Command.XYZ_POS
        self.init_x = rospy.get_param("/" + node_name + "/init_x")
        self.init_y = rospy.get_param("/" + node_name + "/init_y")
        self.init_z = rospy.get_param("/" + node_name + "/init_z")
        self.init_R = rospy.get_param("/" + node_name + "/init_R")
        self.init_P = rospy.get_param("/" + node_name + "/init_P")
        self.init_Y = rospy.get_param("/" + node_name + "/init_Y")
        rospy.Subscriber(topic_header + "local_position/pose", PoseStamped, self.pos_cb, queue_size=20)
        rospy.Subscriber(topic_header + "local_position/velocity_local", TwistStamped, self.vel_cb, queue_size=20)
        self.ext_cmd_pub = rospy.Publisher("/" + node_name + "/px4_cmd/external_command", Command, queue_size=50)
        while rospy.is_shutdown():
            time.sleep(self.update_time)
        progress_1 = threading.Thread(target=self.ros_sub_thread)
        progress_1.start()
        progress_2 = threading.Thread(target=self.ros_pub_thread)
        progress_2.start()

    def ros_sub_thread(self):
        """
        subscribe thread function
        """
        rospy.spin()

    def ros_pub_thread(self):
        """
        publish thread function
        """
        while not rospy.is_shutdown():
            self.ext_cmd_pub.publish(self.external_cmd)
            time.sleep(self.update_time)

    def pos_cb(self, msg: PoseStamped):
        """
        position and pose subscriber callback function
        """
        self.pos_cb_mutex.acquire()
        self.position[0] = msg.pose.position.x + self.init_x
        self.position[1] = msg.pose.position.y + self.init_y
        self.position[2] = msg.pose.position.z + self.init_z
        self.quaternion = [msg.pose.orientation.x, msg.pose.orientation.y, msg.pose.orientation.z, msg.pose.orientation.w]
        (R, P, Y) = euler_from_quaternion(self.quaternion)
        self.attitude[0] = P + self.init_P
        self.attitude[1] = R + self.init_R
        self.attitude[2] = Y + self.init_Y
        self.pos_cb_mutex.release()

    def vel_cb(self, msg: TwistStamped):
        """
        velocity and angle rate subscriber callback function
        """
        self.vel_cb_mutex.acquire()
        self.velocity[0] = msg.twist.linear.x
        self.velocity[1] = msg.twist.linear.y
        self.velocity[2] = msg.twist.linear.z
        self.angle_rate[0] = msg.twist.angular.x
        self.angle_rate[1] = msg.twist.angular.y
        self.angle_rate[2] = msg.twist.angular.z
        self.vel_cb_mutex.release()

    def set_position(self, x: float, y: float, z: float, yaw: Optional[float] = None, yaw_rate: bool = False, frame: int = Command.ENU):
        """setting position command in 3 axis for vehicle

        Args:
            x (float): desire position in x axis, m
            y (float): desire position in y axis, m
            z (float): desire position in z axis, m
            yaw (float | None, optional): desire yaw (rad) / yaw_rate (rad/s) command. Defaults to None.
            yaw_rate (bool): bool flag, if true means yaw rate command, false (default) means yaw command.
            frame (int, optional): position in which frame, px4_cmd.Command.ENU / px4_cmd.Command.BODY. Defaults to Command.ENU.
        """
        self.change_cmd_mutex.acquire()
        self.external_cmd.Mode = Command.Move
        self.external_cmd.Move_frame = frame
        self.external_cmd.Move_mode = Command.XYZ_POS
        self.external_cmd.desire_cmd[0] = x
        self.external_cmd.desire_cmd[1] = y
        self.external_cmd.desire_cmd[2] = z
        self.external_cmd.yaw_cmd = NaN
        self.external_cmd.yaw_rate_cmd = NaN
        if yaw != None:
            if yaw_rate:
                self.external_cmd.yaw_rate_cmd = yaw
            else:
                self.external_cmd.yaw_cmd = yaw
        self.change_cmd_mutex.release()
    
    def set_velocity(self, vx: float, vy: float, vz: float, yaw: Optional[float] = None, yaw_rate: bool = False, frame: int = Command.ENU):
        """setting velocity command in 3 axis for vehicle

        Args:
            vx (float): desire velocity in x axis, m/s
            vy (float): desire velocity in y axis, m/s
            vz (float): desire velocity in z axis, m/s
            yaw (float | None, optional): desire yaw (rad) / yaw_rate (rad/s) command. Defaults to None.
            yaw_rate (bool): bool flag, if true means yaw rate command, false (default) means yaw command.
            frame (int, optional): velocity in which frame, px4_cmd.Command.ENU / px4_cmd.Command.BODY. Defaults to Command.ENU.
        """
        self.change_cmd_mutex.acquire()
        self.external_cmd.Mode = Command.Move
        self.external_cmd.Move_frame = frame
        self.external_cmd.Move_mode = Command.XYZ_VEL
        self.external_cmd.desire_cmd[0] = vx
        self.external_cmd.desire_cmd[1] = vy
        self.external_cmd.desire_cmd[2] = vz
        self.external_cmd.yaw_cmd = NaN
        self.external_cmd.yaw_rate_cmd = NaN
        if yaw != None:
            if yaw_rate:
                self.external_cmd.yaw_rate_cmd = yaw
            else:
                self.external_cmd.yaw_cmd = yaw
        self.change_cmd_mutex.release()

    def set_velocity_with_height(self, vx: float, vy: float, z: float, yaw: Optional[float] = None, yaw_rate: bool = False, frame: int = Command.ENU):
        """setting velocity command in 2 axis with height command for vehicle

        Args:
            vx (float): desire velocity in x axis, m/s
            vy (float): desire velocity in y axis, m/s
            z (float): desire height, m
            yaw (float | None, optional): desire yaw (rad) / yaw_rate (rad/s) command. Defaults to None.
            yaw_rate (bool): bool flag, if true means yaw rate command, false (default) means yaw command.
            frame (int, optional): velocity in which frame, px4_cmd.Command.ENU / px4_cmd.Command.BODY. Defaults to Command.ENU.
        """
        self.change_cmd_mutex.acquire()
        self.external_cmd.Mode = Command.Move
        self.external_cmd.Move_frame = frame
        self.external_cmd.Move_mode = Command.XY_VEL_Z_POS
        self.external_cmd.desire_cmd[0] = vx
        self.external_cmd.desire_cmd[1] = vy
        self.external_cmd.desire_cmd[2] = z
        self.external_cmd.yaw_cmd = NaN
        self.external_cmd.yaw_rate_cmd = NaN
        if yaw != None:
            if yaw_rate:
                self.external_cmd.yaw_rate_cmd = yaw
            else:
                self.external_cmd.yaw_cmd = yaw
        self.change_cmd_mutex.release()
    
    def set_custom_command(self, cmd: CustomCommand) -> None:
        """setting custom command

        Args:
            cmd (CustomCommand): custom command struct
        """        
        self.change_cmd_mutex.acquire()
        self.external_cmd = custom_command_to_px4_msg(cmd)
        self.change_cmd_mutex.release()

    def set_hover(self, yaw: Optional[float] = None) -> None:
        """setting vehicle to hover mode

        Args:
            yaw (float | None, optional): desire yaw command. Defaults to None.
        """
        self.change_cmd_mutex.acquire()
        self.external_cmd.Mode = Command.Hover
        self.external_cmd.Move_frame = Command.ENU
        self.external_cmd.Move_mode = Command.XY_VEL_Z_POS
        self.external_cmd.yaw_cmd = NaN
        self.external_cmd.yaw_rate_cmd = NaN
        if yaw != None:
            self.external_cmd.yaw_cmd = yaw
        self.change_cmd_mutex.release()

File: src/API/vehicle_external_command.cpp
// Copyright (c) 2023 易鹏 中山大学航空航天学院
// Copyright (c) 2023 Peng Yi, Sun Yat-Sen University, School of Aeronautics and Astronautics

#include <px4_cmd/vehicle_external_command.h>

#define PI 3.14159265358979323846
using namespace std;

void vehicle_external_command::start(string node)
{
    string node_name = node;
    string topic_header = "/" + node_name + "/mavros/";
    // handle null string
    if (node_name == "")
    {
        topic_header = "/mavros/";
    }
    external_cmd.Mode = px4_cmd::Command::Hover;
    external_cmd.Move_frame = px4_cmd::Command::ENU;
    external_cmd.Move_mode = px4_cmd::Command::XYZ_POS;
    int argc = 0;
    char **argv;
    ros::init(argc, argv, "ext_cmd");
    ros::NodeHandle nh("/" + node_name, "px4_cmd");
    ros::param::get(("/" + node_name + "/init_x").c_str(), init_x);
    ros::param::get(("/" + node_name + "/init_y").c_str(), init_y);
    ros::param::get(("/" + node_name + "/init_z").c_str(), init_z);
    ros::param::get(("/" + node_name + "/init_R").c_str(), init_R);
    ros::param::get(("/" + node_name + "/init_P").c_str(), init_P);
    ros::param::get(("/" + node_name + "/init_Y").c_str(), init_Y);
    pos_pose_sub = nh.subscribe<geometry_msgs::PoseStamped>(topic_header + "local_position/pose", 20, &vehicle_external_command::pos_cb, this);
    vel_angle_rate_sub = nh.subscribe<geometry_msgs::TwistStamped>(topic_header + "local_position/velocity_local", 20, &vehicle_external_command::vel_cb, this);
    ext_cmd_pub = nh.advertise<px4_cmd::Command>("external_command", 50);
    ext_state_sub = nh.subscribe<std_msgs::Bool>("/" + node_name + "/px4_cmd/ext_cmd_state", 20, &vehicle_external_command::ext_state_cb, this);
    while (!ros::ok())
    {
        usleep(floor(1000000 * update_time));
    }
    std::thread ros_thread(&vehicle_external_command::ros_thread_fun, this);
    ros_thread.detach();
};

void vehicle_external_command::start()
{
    ros::master::V_TopicInfo topics;
    ros::master::getTopics(topics);
    string node_name = "";
    for (auto topic = topics.begin(); topic != topics.end(); topic++)
    {
        auto position = topic->name.find("/mavros");
        if (position != std::string::npos)
        {
            if (position != 0)
            {
                node_name = topic->name.substr(0, position);
            }
            break;
        }
    }
    start(node_name);
};

void vehicle_external_command::ros_thread_fun()
{
    // time counter
    double t = 0;
    // ros main loop
    while (ros::ok() && !ros_shutdown_flag)
    {
        if (ext_cmd_pub.getNumSubscribers() < 1)
        {
            t = 0;
        }
        external_cmd.ext_time = t;
        external_cmd.ext_total_time = total_time;
        ext_cmd_pub.publish(external_cmd);
        t = t + update_time;
        usleep(floor(1000000 * update_time));
        ros::spinOnce();
    }
    // exit with shutdown
    ros::shutdown();
};

void vehicle_external_command::pos_cb(const geometry_msgs::PoseStamped::ConstPtr &msg)
{
    pos_cb_mutex.lock();
    position[0] = msg->pose.position.x + init_x;
    position[1] = msg->pose.position.y + init_y;
    position[2] = msg->pose.position.z + init_z;
    double R;
    double P;
    double Y;
    tf::quaternionMsgToTF(msg->pose.orientation, quaternion);
    tf::Matrix3x3(quaternion).getRPY(R, P, Y);
    attitude[0] = P + init_P;
    attitude[1] = R + init_R;
    attitude[2] = Y + init_Y;
    pos_cb_mutex.unlock();
};

void vehicle_external_command::vel_cb(const geometry_msgs::TwistStamped::ConstPtr &msg)
{
    vel_cb_mutex.lock();
    velocity[0] = msg->twist.linear.x;
    velocity[1] = msg->twist.linear.y;
    velocity[2] = msg->twist.linear.z;
    angle_rate[0] = msg->twist.angular.x * 180 / PI;
    angle_rate[1] = msg->twist.angular.y * 180 / PI;
    angle_rate[2] = msg->twist.angular.z * 180 / PI;
    vel_cb_mutex.unlock();
};

void vehicle_external_command::ext_state_cb(const std_msgs::Bool::ConstPtr &msg)
{
    ext_cmd_state = msg->data;
};

void vehicle_external_command::set_position_thread_func(double x, double y, double z, int frame)
{
    while (check_sync_lock())
    {
        usleep(50);
    }
    change_cmd_mutex.lock();
    external_cmd.Mode = px4_cmd::Command::Move;
    external_cmd.Move_frame = frame;
    external_cmd.Move_mode = px4_cmd::Command::XYZ_POS;
    external_cmd.desire_cmd[0] = x;
    external_cmd.desire_cmd[1] = y;
    external_cmd.desire_cmd[2] = z;
    external_cmd.yaw_cmd = NAN;
    external_cmd.yaw_rate_cmd = NAN;
    change_cmd_mutex.unlock();
};

void vehicle_external_command::set_position_thread_func(double x, double y, double z, double yaw_cmd, bool yaw_rate, int frame)
{
    while (check_sync_lock())
    {
        usleep(50);
    }
    change_cmd_mutex.lock();
    external_cmd.Mode = px4_cmd::Command::Move;
    external_cmd.Move_frame = frame;
    external_cmd.Move_mode = px4_cmd::Command::XYZ_POS;
    external_cmd.desire_cmd[0] = x;
    external_cmd.desire_cmd[1] = y;
    external_cmd.desire_cmd[2] = z;
    if (yaw_rate)
    {
        external_cmd.yaw_cmd = NAN;
        external_cmd.yaw_rate_cmd = yaw_cmd;
    }
    else
    {
        external_cmd.yaw_cmd = yaw_cmd - init_Y;
        external_cmd.yaw_rate_cmd = NAN;
    }
    change_cmd_mutex.unlock();
};

void vehicle_external_command::set_velocity_thread_func(double vx, double vy, double vz, int frame)
{
    while (check_sync_lock())
    {
        usleep(50);
    }
    change_cmd_mutex.lock();
    external_cmd.Mode = px4_cmd::Command::Move;
    external_cmd.Move_frame = frame;
    external_cmd.Move_mode = px4_cmd::Command::XYZ_VEL;
    external_cmd.desire_cmd[0] = vx;
    external_cmd.desire_cmd[1] = vy;
    external_cmd.desire_cmd[2] = vz;
    external_cmd.yaw_cmd = NAN;
    external_cmd.yaw_rate_cmd = NAN;
    change_cmd_mutex.unlock();
};

void vehicle_external_command::set_velocity_thread_func(double vx, double vy, double vz, double yaw_cmd, bool yaw_rate, int frame)
{
    while (check_sync_lock())
    {
        usleep(50);
    }
    change_cmd_mutex.lock();
    external_cmd.Mode = px4_cmd::Command::Move;
    external_cmd.Move_frame = frame;
    external_cmd.Move_mode = px4_cmd::Command::XYZ_VEL;
    external_cmd.desire_cmd[0] = vx;
    external_cmd.desire_cmd[1] = vy;
    external_cmd.desire_cmd[2] = vz;
    if (yaw_rate)
    {
        external_cmd.yaw_cmd = NAN;
        external_cmd.yaw_rate_cmd = yaw_cmd;
    }
    else
    {
        external_cmd.yaw_cmd = yaw_cmd - init_Y;
        external_cmd.yaw_rate_cmd = NAN;
    }
    change_cmd_mutex.unlock();
};

void vehicle_external_command::set_velocity_with_height_thread_func(double vx, double vy, double z, int frame)
{
    while (check_sync_lock())
    {
        usleep(50);
    }
    change_cmd_mutex.lock();
    external_cmd.Mode = px4_cmd::Command::Move;
    external_cmd.Move_frame = frame;
    external_cmd.Move_mode = px4_cmd::Command::XY_VEL_Z_POS;
    external_cmd.desire_cmd[0] = vx;
    external_cmd.desire_cmd[1] = vy;
    external_cmd.desire_cmd[2] = z;
    external_cmd.yaw_cmd = NAN;
    external_cmd.yaw_rate_cmd = NAN;
    change_cmd_mutex.unlock();
};

void vehicle_external_command::set_velocity_with_height_thread_func(double vx, double vy, double z, double yaw_cmd, bool yaw_rate, int frame)
{
    while (check_sync_lock())
    {
        usleep(50);
    }
    change_cmd_mutex.lock();
    external_cmd.Mode = px4_cmd::Command::Move;
    external_cmd.Move_frame = frame;
    external_cmd.Move_mode = px4_cmd::Command::XY_VEL_Z_POS;
    external_cmd.desire_cmd[0] = vx;
    external_cmd.desire_cmd[1] = vy;
    external_cmd.desire_cmd[2] = z;
    if (yaw_rate)
    {
        external_cmd.yaw_cmd = NAN;
        external_cmd.yaw_rate_cmd = yaw_cmd;
    }
    else
    {
        external_cmd.yaw_cmd = yaw_cmd - init_Y;
        external_cmd.yaw_rate_cmd = NAN;
    }
    change_cmd_mutex.unlock();
};

void vehicle_external_command::set_custom_command_thread_func(CustomCommand cmd)
{
    while (check_sync_lock())
    {
        usleep(50);
    }
    change_cmd_mutex.lock();
    custom_command_to_px4_msg(cmd, external_cmd);
    change_cmd_mutex.unlock();
}

void vehicle_external_command::set_hover_thread_func()
{
    while (check_sync_lock())
    {
        usleep(50);
    }
    change_cmd_mutex.lock();
    external_cmd.Mode = px4_cmd::Command::Hover;
    external_cmd.Move_frame = px4_cmd::Command::ENU;
    external_cmd.Move_mode = px4_cmd::Command::XYZ_POS;
    external_cmd.yaw_cmd = NAN;
    external_cmd.yaw_rate_cmd = NAN;
    change_cmd_mutex.unlock();
};

void vehicle_external_command::set_hover_thread_func(double yaw)
{
    while (check_sync_lock())
    {
        usleep(50);
    }
    change_cmd_mutex.lock();
    external_cmd.Mode = px4_cmd::Command::Hover;
    external_cmd.Move_frame = px4_cmd::Command::ENU;
    external_cmd.Move_mode = px4_cmd::Command::XYZ_POS;
    external_cmd.yaw_cmd = yaw;
    external_cmd.yaw_rate_cmd = NAN;
    change_cmd_mutex.unlock();
};

void vehicle_external_command::set_position(double x, double y, double z, int frame)
{
    void (vehicle_external_command::*thread_func)(double, double, double, int) = &vehicle_external_command::set_position_thread_func;
    std::thread set_cmd_thread(std::bind(thread_func, this, x, y, z, frame));
    set_cmd_thread.detach();
};

void vehicle_external_command::set_position(double x, double y, double z, double yaw_cmd, bool yaw_rate, int frame)
{
    void (vehicle_external_command::*thread_func)(double, double, double, double, bool, int) = &vehicle_external_command::set_position_thread_func;
    std::thread set_cmd_thread(std::bind(thread_func, this, x, y, z, yaw_cmd, yaw_rate, frame));
    set_cmd_thread.detach();
};

void vehicle_external_command::set_velocity(double vx, double vy, double vz, int frame)
{
    void (vehicle_external_command::*thread_func)(double, double, double, int) = &vehicle_external_command::set_velocity_thread_func;
    std::thread set_cmd_thread(std::bind(thread_func, this, vx, vy, vz, frame));
    set_cmd_thread.detach();
};

void vehicle_external_command::set_velocity(double vx, double vy, double vz, double yaw_cmd, bool yaw_rate, int frame)
{
    void (vehicle_external_command::*thread_func)(double, double, double, double, bool, int) = &vehicle_external_command::set_velocity_thread_func;
    std::thread set_cmd_thread(std::bind(thread_func, this, vx, vy, vz, yaw_cmd, yaw_rate, frame));
    set_cmd_thread.detach();
};

void vehicle_external_command::set_velocity_with_height(double vx, double vy, double z, int frame)
{
    void (vehicle_external_command::*thread_func)(double, double, double, int) = &vehicle_external_command::set_velocity_with_height_thread_func;
    std::thread set_cmd_thread(std::bind(thread_func, this, vx, vy, z, frame));
    set_cmd_thread.detach();
};

void vehicle_external_command::set_velocity_with_height(double vx, double vy, double z, double yaw_cmd, bool yaw_rate, int frame)
{
    void (vehicle_external_command::*thread_func)(double, double, double, double, bool, int) = &vehicle_external_command::set_velocity_with_height_thread_func;
    std::thread set_cmd_thread(std::bind(thread_func, this, vx, vy, z, yaw_cmd, yaw_rate, frame));
    set_cmd_thread.detach();
};

void vehicle_external_command::set_custom_command(CustomCommand cmd)
{
    std::thread set_cmd_thread(&vehicle_external_command::set_custom_command_thread_func, this, cmd);
    set_cmd_thread.detach();
}

void vehicle_external_command::set_hover()
{
    void (vehicle_external_command::*thread_func)() = &vehicle_external_command::set_hover_thread_func;
    std::thread set_cmd_thread(std::bind(thread_func, this));
    set_cmd_thread.detach();
};

void vehicle_external_command::set_hover(double yaw)
{
    void (vehicle_external_command::*thread_func)(double) = &vehicle_external_command::set_hover_thread_func;
    std::thread set_cmd_thread(std::bind(thread_func, this, yaw));
    set_cmd_thread.detach();
};

bool vehicle_external_command::check_sync_lock()
{
    if (sync_lock == nullptr)
    {
        return false;
    }
    return *sync_lock;
};

void vehicle_external_command::set_sync_lock(bool *sync_lock_)
{
    sync_lock = sync_lock_;
};

void vehicle_external_command::reset_sync_lock()
{
    sync_lock = nullptr;
};

void vehicle_external_command::shutdown()
{
    ros_shutdown_flag = true;
};